File: apps/demo-be-fsharp-giraffe/tests/DemoBeFsgi.Tests/Integration/Steps/SecuritySteps.fs
module DemoBeFsgi.Tests.Integration.Steps.SecuritySteps

open TickSpec
open Xunit
open DemoBeFsgi.Tests.State
open DemoBeFsgi.Tests.Integration.Steps.CommonSteps
open DemoBeFsgi.Tests.Integration.Steps.TokenManagementSteps

[<Given>]
let ``"(.+)" has had the maximum number of failed login attempts`` (username: string) (state: StepState) =
    // Send 5 failed login attempts with a wrong password
    for _ in 1..5 do
        sendPost
            state.Client
            "/api/v1/auth/login"
            $"""{{ "username": "{username}", "password": "WrongPass1234!" }}"""
            None
        |> ignore

    // Get alice's user ID via admin
    registerUser state.Client "tempAdmin_fail" "tempAdmin_fail@example.com" "Str0ng#Admin1"
    |> ignore

    sendPost state.Client "/test/set-admin-role/tempAdmin_fail" "" None |> ignore
    let adminToken, _ = loginUser state.Client "tempAdmin_fail" "Str0ng#Admin1"
    let email = $"{username}@example.com"

    let response, body =
        sendGet state.Client $"/api/v1/admin/users?email={email}" adminToken

    let doc = System.Text.Json.JsonDocument.Parse(body)
    let dataEl = doc.RootElement.GetProperty("data")
    let userId = ref ""

    for item in dataEl.EnumerateArray() do
        try
            let idEl = item.GetProperty("id")
            userId.Value <- idEl.GetString()
        with _ ->
            ()

    { state with
        UserId = Some userId.Value }

[<Then>]
let ``alice's account status should be "(.+)"`` (expectedStatus: string) (state: StepState) =
    // We need to check the account status. We can try to call an admin endpoint or
    // check the error message from login.
    // For "locked": a locked account returns 401 with locked message.
    // The response we have should indicate locked status.
    let body = state.ResponseBody.Value

    match expectedStatus with
    | "locked" ->
        // The response should have an error about locked account or we check status via admin
        Assert.True(
            body.ToLower().Contains("lock") || body.ToLower().Contains("attempt"),
            $"Expected locked account message but got: {body}"
        )
    | "disabled" ->
        Assert.True(
            body.ToLower().Contains("disabl") || int state.Response.Value.StatusCode = 401,
            $"Expected disabled account message but got: {body}"
        )
    | "active" ->
        // Check that we got a successful response or token
        let statusCode = int state.Response.Value.StatusCode

        Assert.True(
            statusCode = 200 || statusCode = 201,
            $"Expected active account (200/201) but got {statusCode}: {body}"
        )
    | _ -> Assert.True(body.Length > 0)

    state

[<Given>]
let ``a user "(.+)" is registered and locked after too many failed logins`` (username: string) (state: StepState) =
    let email = $"{username}@example.com"
    registerUser state.Client username email "Str0ng#Pass1" |> ignore

    // Trigger lockout with wrong password
    for _ in 1..5 do
        sendPost
            state.Client
            "/api/v1/auth/login"
            $"""{{ "username": "{username}", "password": "WrongPass123!" }}"""
            None
        |> ignore

    // Get alice's ID by registering a temp admin and listing users
    registerUser state.Client "tempAdmin_lock" "tempAdmin_lock@example.com" "Str0ng#Admin1"
    |> ignore

    sendPost state.Client "/test/set-admin-role/tempAdmin_lock" "" None |> ignore
    let adminToken, _ = loginUser state.Client "tempAdmin_lock" "Str0ng#Admin1"

    let response, body =
        sendGet state.Client $"/api/v1/admin/users?email={email}" adminToken

    let doc = System.Text.Json.JsonDocument.Parse(body)
    let dataEl = doc.RootElement.GetProperty("data")
    let aliceId = ref ""

    for item in dataEl.EnumerateArray() do
        try
            let idEl = item.GetProperty("id")
            aliceId.Value <- idEl.GetString()
        with _ ->
            ()

    { state with
        UserId = Some aliceId.Value }

[<Given>]
let ``an admin has unlocked alice's account`` (state: StepState) =
    // Register admin and unlock alice
    registerUser state.Client "testadmin_sec" "testadmin_sec@example.com" "Str0ng#Admin1"
    |> ignore

    sendPost state.Client "/test/set-admin-role/testadmin_sec" "" None |> ignore
    let adminToken, _ = loginUser state.Client "testadmin_sec" "Str0ng#Admin1"

    // Find alice's user ID
    let email = "alice@example.com"

    let response, body =
        sendGet state.Client $"/api/v1/admin/users?email={email}" adminToken

    let doc = System.Text.Json.JsonDocument.Parse(body)
    let dataEl = doc.RootElement.GetProperty("data")
    let aliceId = ref ""

    for item in dataEl.EnumerateArray() do
        try
            let idEl = item.GetProperty("id")
            aliceId.Value <- idEl.GetString()
        with _ ->
            ()

    if aliceId.Value <> "" then
        sendPost state.Client $"/api/v1/admin/users/{aliceId.Value}/unlock" "" adminToken
        |> ignore

    { state with
        UserId = Some aliceId.Value }

[<When>]
let ``the admin sends POST /api/v1/admin/users/\{alice_id\}/unlock`` (state: StepState) =
    let adminToken = state.ExtraData |> Map.tryFind "adminToken"
    let aliceId = state.UserId |> Option.defaultValue ""

    let response, responseBody =
        sendPost state.Client $"/api/v1/admin/users/{aliceId}/unlock" "" adminToken

    { state with
        Response = Some response
        ResponseBody = Some responseBody }

File: apps/demo-be-python-fastapi/tests/unit/steps/security_steps.py
"""Unit BDD step definitions for security feature."""

import pytest
from fastapi.testclient import TestClient
from pytest_bdd import given, parsers, scenarios, then, when

from demo_be_python_fastapi.config import settings
from tests.unit.conftest import GHERKIN_ROOT
from tests.unit.steps._helpers import _ADMIN_PASSWORD, register_and_promote_admin

pytestmark = pytest.mark.unit

scenarios(str(GHERKIN_ROOT / "security" / "security.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    parsers.parse('a user "{username}" is registered and locked after too many failed logins'),
    target_fixture="locked_user",
)
def register_and_lock_user(client: TestClient, username: str) -> dict:
    resp = client.post(
        "/api/v1/auth/register",
        json={
            "username": username,
            "email": f"{username}@example.com",
            "password": _PASSWORD,
        },
    )
    assert resp.status_code == 201
    user_data = resp.json()

    for _ in range(settings.max_failed_login_attempts):
        client.post(
            "/api/v1/auth/login",
            json={"username": username, "password": "WrongPass#1234"},
        )
    return user_data


@given(
    parsers.parse('an admin user "{username}" is registered and logged in'),
    target_fixture="admin_tokens",
)
def register_admin_and_login(client: TestClient, username: str) -> dict:
    user_data = register_and_promote_admin(client, username, _ADMIN_PASSWORD)
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": username, "password": _ADMIN_PASSWORD},
    )
    assert resp.status_code == 200, f"Admin login failed: {resp.text}"
    return {**resp.json(), "id": user_data["id"]}


@given(
    parsers.parse('"alice" has had the maximum number of failed login attempts'),
)
def alice_max_failed_attempts(client: TestClient, registered_user: dict) -> None:
    for _ in range(settings.max_failed_login_attempts):
        client.post(
            "/api/v1/auth/login",
            json={"username": "alice", "password": "WrongPass#1234"},
        )


@given("an admin has unlocked alice's account")
def admin_unlocks_alice(client: TestClient, locked_user: dict) -> None:
    register_and_promote_admin(client, "tmpadmin", _ADMIN_PASSWORD)
    login_resp = client.post(
        "/api/v1/auth/login",
        json={"username": "tmpadmin", "password": _ADMIN_PASSWORD},
    )
    assert login_resp.status_code == 200, f"Admin login failed: {login_resp.text}"
    tmp_tokens = login_resp.json()
    resp = client.post(
        f"/api/v1/admin/users/{locked_user['id']}/unlock",
        headers={"Authorization": f"Bearer {tmp_tokens['access_token']}"},
    )
    assert resp.status_code == 200, f"Unlock failed: {resp.text}"


# --- When steps ---


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/unlock"),
    target_fixture="response",
)
def admin_unlock_alice(client: TestClient, locked_user: dict, admin_tokens: dict):  # type: ignore[no-untyped-def]
    return client.post(
        f"/api/v1/admin/users/{locked_user['id']}/unlock",
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


# --- Then steps ---


@then('alice\'s account status should be "locked"')
def check_alice_locked(client: TestClient, registered_user: dict) -> None:
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": "alice", "password": _PASSWORD},
    )
    assert resp.status_code == 401, (
        f"Expected 401 for locked account, got {resp.status_code}: {resp.text}"
    )
